engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java

```diff
@@ -3516,6 +3517,16 @@ private void checkPlatformCompatible(AppPackage ap)
     }
   }
 
+  private void checkJavaVersionCompatible(AppPackage ap)
+  {
+    String buildJdkVersion = ap.getBuildJdkVersion();
+    String systemVersion = System.getProperty("java.version");
+    if (VersionInfo.compare(buildJdkVersion, systemVersion) > 0) {
```

Comment:
I don't see why checking every class is necessary. An application package may contain classes that are not used or used only if JVM version supports them. Additionally Apex client JVM may not match JVMs on the cluster. It should be sufficient to catch UnsupportedClassVersionError.